### PR TITLE
eve ouput kafka v2

### PR DESCRIPTION
**Pull request by fooinha**
Initial work for logging output into the distributed streaming platform Apache Kafka.

Allows to send eve messages to a list of brokers of Apache Kafka.
It uses the client implementation from [librdkafka](https://github.com/edenhill/librdkafka).

It's set to be conditional at configure/compile time with the  --enable-rdkafka [Enable Kafka support] option.

Configuration settings and default values: 

``` yaml
      filetype: regular #regular|syslog|unix_dgram|unix_stream|redis|kafka
       filename: eve.json
       kafka:
        broker-list: 127.0.0.1:9092
        topic: suricata
        compression: snappy # none, gzip, snappy
        partition: -1 # below < 0 means Unassigned Partition
        max-retries: 1
        backoff-ms: 10
        buffer-max-messages: 100000
        log-level: 6
```

It's an alternative for redis output. 

Rebased to upstream/master HEAD.

PPA for precise build included at travis.yaml

### CODE REVIEW

**Comment by inliniac** [comment]
Passes my QA. This merely means that it's not breaking anything if librdkafka is not installed.

Can you squash https://github.com/inliniac/suricata/pull/2338/commits/ef36554cafb4cba5b8bd2d2a3e44a793c73a2cad ? We don't want commits fixing other commits in a single PR.

Also, can you merge the doc branch into this one, while keeping the doc commit separate from the code commit.

@jasonish are you interested in looking at the code & functionality?

**Comment by fooinha** [comment]
If you want, for testing, I can fork from amsterdam project and add kafka brokers requirements to docker.

**Comment by inliniac** [comment]
On 07-10-16 15:06, fooinha wrote:
> If you want, for testing, I can fork from amsterdam project and add
> kafka brokers requirements to docker.

I think that would be pretty cool!

File: src/util-logopenfile.c [hunk 1]
```diff
@@ -465,6 +469,336 @@ int SCConfLogReopenRedis(LogFileCtx *log_ctx)
...
+        exit(EXIT_FAILURE);
+    }
+
+    if (partition < 0) {
+        partition = kafka_default_partition;
+        SCLogInfo("eve kafka output: using default unassigned partition");
+    }
+
+    /* Configures and starts up kafka things */
+    {
```
Comment (jasonish) [inline src/util-logopenfile.c]: Whenever I see this I think it could be refactored a bit.

**Comment by jasonish** [comment]
> @jasonish are you interested in looking at the code & functionality?

I looked enough to make sure the logging call would never block (I believe the Redis output can block, and IMO, we don't want anything that will block - anymore than writing a file).

I'm a little curious what happens if the kafka server can't keep up? Does the kafka lib just keep queuing up messages and consuming memory? Or will it drop messages?  I don't think these cases necessarily need to be handled gracefully, but at least known?

**Comment by fooinha** [comment]
I think that with this version of librdkafka (0.8.6) if queue reaches the queue.buffering.max.messages conf value / buffer-max-messages: 2048 in suricata.yaml , it just drops the message.

In the following versions there is a msg flag to block (default behaviour) producer thread queue and retry if queue is full.

note: the build with clang failed because of missing package dep. the gcc did the same and it worked  😕

**Comment by fooinha** [comment]
I've changed the max messages buffer to match librdkafka defauls. 100k.

**Comment by fooinha** [comment]
Created a repository with Dockerfile that creates images for a suricata 3.2.0beta1 with experimental version with eve kafka output.

It is based on stamus/suricata docker image.
The stamus/suricata uses debian jessie (stable) and suricata 2.x versions.

As we want to test this, I upgraded derived image to sid (unstable) and installed  suricata 3.2.0beta1 compiled with rdkafka support.

Dockerfile project is here -> https://github.com/fooinha/docker-stamus-suricata-kafka 

Also, forked the Amsterdam project to include this suricata image and added to docker-compose the zookeeper and kafka brokers images and links.

The fork with updated configuration for suricata and logstash and with new services is in this branch -> https://github.com/fooinha/Amsterdam/tree/feature/suricata/kafka

Some things related with rules configuration,  don't work, as i didn't bother to do it, but we already can see event messages being consumed from kafka to elasticsearch.

Note: sometimes it may exist a race condition  between zookeeper and kafka containers and kafka broker may forfeit and shutdown. For now, simple Ctrl-c and restart it.

**Comment by fooinha** [comment]
Recover from suspend state works nicely! kewl!

```c
elasticsearch_1  | [2016-10-11 07:32:37,070][INFO ][cluster.metadata         ] [Amphibian] [logstash-http-2016.10.11] update_mapping [SELKS]
...
...
...
kafka_1          | [2016-10-11 19:38:38,467] INFO Client session timed out, have not heard from server in 43412895ms for sessionid 0x157af446b9b0002, closing socket connection and attempting reconnect (org.apache.zookeeper.ClientCnxn)
zookeeper_1      | 2016-10-11 19:38:38,470 [myid:] - INFO  [SessionTracker:ZooKeeperServer@358] - Expiring session 0x157af446b9b0003, timeout of 6000ms exceeded
...
...
...
suricata_1       | [18] 11/10/2016 -- 19:38:38 - (source-af-packet.c:1412) <Error> (ReceiveAFPLoop) -- [ERRCODE: SC_ERR_AFP_READ(191)] - Error reading data from iface 'eth0': (100u) Network is down
kafka_1          | [2016-10-11 19:38:40,033] INFO Result of znode creation is: OK (kafka.utils.ZKCheckedEphemeral)
kafka_1          | [2016-10-11 19:38:40,033] INFO Registered broker 0 at path /brokers/ids/0 with addresses: PLAINTEXT -> EndPoint(kafka,9092,PLAINTEXT) (kafka.utils.ZkUtils)
kafka_1          | [2016-10-11 19:38:40,033] INFO done re-registering broker (kafka.server.KafkaHealthcheck$SessionExpireListener)
kafka_1          | [2016-10-11 19:38:40,033] INFO Subscribing to /brokers/topics path to watch for new topics (kafka.server.KafkaHealthcheck$SessionExpireListener)
kafka_1          | [2016-10-11 19:38:40,069] INFO New leader is 0 
elasticsearch_1  | [2016-10-11 19:43:51,856][INFO ][cluster.metadata         ] [Amphibian] [logstash-fileinfo-2016.10.11] update_mapping [SELKS]
elasticsearch_1  | [2016-10-11 19:43:51,871][INFO ][cluster.metadata         ] [Amphibian] [logstash-http-2016.10.11] update_mapping [SELKS]
logstash_1       | EVE insert rate: 0.5027287813153638

```

```c

T 2016/10/11 19:44:00.679670 10.0.10.1:41940 -> 10.0.10.6:9092 [AP] ..............atom............suricata.........................8.u..........
{"timestamp":"2016-10-11T19:44:00.648568+0000","flow_id":1737201544469305,
"in_iface":"eth0","event_type":"fileinfo","src_ip":"93.184.220.29","src_port":80,
"dest_ip":"192.168.1.84","dest_port":33644,"proto":"TCP","http":{"hostname":"ocsp.digicert.com","url":"\/",
"http_user_agent":"Mozilla\/5.0 
(X11; Ubuntu; Linux x86_64; rv:49.0) Gecko\/20100101 Firefox\/49.0",
"http_content_type":"application\/ocsp-response","http_method":"POST",
"protocol":"HTTP\/1.1","status":200,"length":471},"app_proto":"http",
"fileinfo":{"filename":"\/","magic":"data","state":"CLOSED","md5":"95821e628f772f1f76f9bced231f2de6",
"stored":false,"size":471,"tx_id":0},"host":"atom"}

```

**Comment by fooinha** [comment]
I did some work from redis async with #2362

**Comment by driverpt** [comment]
Do we have any ETA on this ?

**Comment by inliniac** [comment]
Replaced by #2436

**Pull request closed without merging**